src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.ca.plugins;
+
+import com.mysql.cj.conf.HostInfo;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.jdbc.ConnectionImpl;
+import com.mysql.cj.log.Log;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class Monitor implements IMonitor {
+  private static final int THREAD_SLEEP_WHEN_INACTIVE_MILLIS = 100;
+
+  private final Queue<MonitorConnectionContext> contexts = new ConcurrentLinkedQueue<>();
+
+  private final Log log;
+  private final PropertySet propertySet;
+  private final HostInfo hostInfo;
+  private boolean isMonitoring;
+  private long monitoringStartTime;
+  private Connection monitoringConn = null;
+  private int longestFailureDetectionIntervalMillis;
+  private final AtomicLong elapsedTime = new AtomicLong();
+
+  public Monitor(HostInfo hostInfo, PropertySet propertySet, Log log) {
+    this.hostInfo = hostInfo;
+    this.propertySet = propertySet;
+    this.log = log;
+    this.isMonitoring = false;
+  }
+
+  @Override
+  public void startMonitoring(MonitorConnectionContext context) {
+    this.longestFailureDetectionIntervalMillis = Math.min(
+        this.longestFailureDetectionIntervalMillis,
+        context.getFailureDetectionIntervalMillis());
+
+    this.monitoringStartTime = System.currentTimeMillis();
+    contexts.add(context);
+  }
+
+  @Override
+  public void stopMonitoring(MonitorConnectionContext context) {
+    this.longestFailureDetectionIntervalMillis = findLongestIntervalMillis();
+    contexts.remove(context);
+    this.isMonitoring = false;
+  }
+
+  @Override
+  public void run() {
+    try {
+      while (true) {
+        if (!contexts.isEmpty() || !isConnectionHealthy()) {
+          for (MonitorConnectionContext monitorContext : contexts) {
+            final long elapsedTimeMillis = System.currentTimeMillis() - this.monitoringStartTime;
```

Comment:
this.monitoringStartTime is wrong. It should be something monitorContext.monitoringStartTime